### PR TITLE
[SPARK-24369][SQL] Correct handling for multiple distinct aggregations having the same argument set

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr fixed an issue when having multiple distinct aggregations having the same argument set, e.g.,
```
scala>: paste
val df = sql(
  s"""SELECT corr(DISTINCT x, y), corr(DISTINCT y, x), count(*)
     | FROM (VALUES (1, 1), (2, 2), (2, 2)) t(x, y)
   """.stripMargin)

java.lang.RuntimeException
You hit a query analyzer bug. Please report your query to Spark user mailing list.
```
The root cause is that `RewriteDistinctAggregates` can't detect multiple distinct aggregations if they have the same argument set. This pr modified code so that `RewriteDistinctAggregates` could count the number of aggregate expressions with `isDistinct=true`.

## How was this patch tested?
Added tests in `DataFrameAggregateSuite`.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala [hunk 1]
```diff
@@ -687,4 +687,12 @@ class DataFrameAggregateSuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("SPARK-24369 multiple distinct aggregations having the same argument set") {
+    val df = sql(
+      s"""SELECT corr(DISTINCT x, y), corr(DISTINCT y, x), count(*)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala]: Move it to SQLQueryTestSuite?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala [hunk 2]
```diff
@@ -151,7 +152,7 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
       }
 
       // Setup unique distinct aggregate children.
-      val distinctAggChildren = distinctAggGroups.keySet.flatten.toSeq.distinct
+      val distinctAggChildren = distinctAggGroups.keySet.flatten.toSeq
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala]: Is this needed?

**Comment by gatorsmile** [comment]
cc @hvanhovell

**Comment by kiszk** [comment]
retest this please

**Comment by cloud-fan** [comment]
good catch! LGTM, merging  to master!

**Comment by gatorsmile** [comment]
For the following query, we could see the performance regression
```SQL
SELECT sum(DISTINCT x), avg(DISTINCT x)
FROM (VALUES (1, 1), (2, 2), (2, 2)) t(x, y)
```

Before this PR:
```
== Optimized Logical Plan ==
Aggregate [sum(distinct cast(x#189 as bigint)) AS sum(DISTINCT x)#193L, avg(distinct cast(x#189 as bigint)) AS avg(DISTINCT x)#194]
+- LocalRelation [x#189]
```

After this PR
```
== Optimized Logical Plan ==
Aggregate [sum(if ((gid#195 = 1)) CAST(`x` AS BIGINT)#196L else null) AS sum(DISTINCT x)#193L, avg(if ((gid#195 = 1)) CAST(`x` AS BIGINT)#196L else null) AS avg(DISTINCT x)#194]
+- Aggregate [CAST(`x` AS BIGINT)#196L, gid#195], [CAST(`x` AS BIGINT)#196L, gid#195]
   +- Expand [List(cast(x#189 as bigint), 1)], [CAST(`x` AS BIGINT)#196L, gid#195]
      +- LocalRelation [x#189]
```

**Comment by gatorsmile** [comment]
I will revert this PR now. @maropu Could you submit a new fix to resolve the above issue?

**Comment by maropu** [comment]
@gatorsmile ok, I wil (so, I reopend https://issues.apache.org/jira/browse/SPARK-24369)

**Pull request closed without merging**